packages/next-swc/crates/next-core/src/app_source.rs

```diff
@@ -99,31 +99,27 @@ use crate::{
     util::{render_data, NextRuntime},
 };
 
-#[turbo_tasks::function]
-fn pathname_to_specificity(pathname: &str) -> SpecificityVc {
-    let mut current = Specificity::new();
-    let mut position = 0;
+fn pathname_to_segments(pathname: &str) -> (Vec<BaseSegment>, Option<FinalSegment>) {
+    let mut segments = Vec::new();
     for segment in pathname.split('/') {
-        if segment.starts_with('(') && segment.ends_with(')') || segment.starts_with('@') {
+        if segment.is_empty()
+            || (segment.starts_with('(') && segment.ends_with(')') || segment.starts_with('@'))
+        {
             // ignore
         } else if segment.starts_with("[[...") && segment.ends_with("]]")
             || segment.starts_with("[...") && segment.ends_with(']')
         {
-            // optional catch all segment
-            current.add(position - 1, SpecificityElementType::CatchAll);
-            position += 1;
-        } else if segment.starts_with("[[") || segment.ends_with("]]") {
-            // optional segment
-            position += 1;
+            // (optional) catch all segment
+            return (segments, Some(FinalSegment::CatchAll));
```

Comment:
Should this asset that the catch all segment is the last segment? Otherwise, we might be discarding invalid segments without warning.